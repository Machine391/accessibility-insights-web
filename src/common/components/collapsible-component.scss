// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

@import '../../common/styles/common.scss';

span.collapsible-title {
    padding-left: 0.5vw;
}

.collapsible {
    display: inline-flex;
    font-size: 26px;
    height: auto;
    width: auto;
    text-align: left;
    color: #106ebe;

    h2 {
        font-weight: normal;
        font-size: 17px;
    }
}

.collapsible-content {
    overflow: hidden;
    padding-left: calc(0.5vw + 13px); // 13px is Chevron width.
}
