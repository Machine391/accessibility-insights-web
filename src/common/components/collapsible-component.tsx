// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { ActionButton } from '@fluentui/react';
import { Button, themeToTokensObject, webLightTheme } from '@fluentui/react-components';
import { css } from '@fluentui/utilities';
import styles from 'common/components/collapsible-component.scss';
import * as React from 'react';
import { ChevronDown24Regular, ChevronRight24Regular } from '@fluentui/react-icons';

export interface CollapsibleComponentProps {
    header: JSX.Element;
    content: JSX.Element;
    contentClassName?: string;
    containerClassName?: string;
}

interface CollapsibleComponentState {
    showContent: boolean;
}

export class CollapsibleComponent extends React.Component<
    CollapsibleComponentProps,
    CollapsibleComponentState
> {
    private readonly iconNameDown = 'ChevronDown';
    private readonly iconNameUp = 'ChevronRight';

    constructor(props: CollapsibleComponentProps) {
        super(props);
        this.state = { showContent: true };
    }

    private onClick = (): void => {
        this.setState(prevState => ({ showContent: !prevState.showContent }));
    };

    public render(): JSX.Element {
        const showContent = this.state.showContent;
        let iconName = this.iconNameUp;
        let content: JSX.Element | null = null;

        if (showContent) {
            iconName = this.iconNameDown;
            content = (
                <div className={css(this.props.contentClassName, styles.collapsibleContent)}>
                    {this.props.content}
                </div>
            );
        }

        return (
            <div className={css(this.props.containerClassName)}>
                {/* 111<ActionButton
                    className={styles.collapsible}
                    iconProps={{ iconName: iconName }}
                    onClick={this.onClick}
                    aria-expanded={showContent}
                >
                    {this.props.header}
                </ActionButton> */}
                <Button className={styles.collapsible} appearance='transparent'
                    onClick={this.onClick}
                    aria-expanded={showContent}>
                    {showContent ? (
                        <ChevronDown24Regular color={themeToTokensObject(webLightTheme).colorCompoundBrandStrokeHover} className={styles.collapsible} />
                    ) : (
                        <ChevronRight24Regular color={themeToTokensObject(webLightTheme).colorCompoundBrandStrokeHover} className={styles.collapsible} />
                    )}
                    111
                    {this.props.header}
                </Button>
                {content}
            </div>
        );
    }
}
