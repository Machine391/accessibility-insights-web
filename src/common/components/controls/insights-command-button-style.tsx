// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { makeStyles, tokens } from '@fluentui/react-components';

export const InsightsCommandButtonStyle = makeStyles({
    button: { fontWeight: tokens.fontWeightRegular },
    buttonIcon: {
        color: tokens.colorCompoundBrandStrokeHover,
    },
});
