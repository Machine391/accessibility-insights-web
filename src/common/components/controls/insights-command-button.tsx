// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { css } from '@fluentui/react';
import { Button, ButtonProps } from '@fluentui/react-components';
import { Icons } from 'common/icons/fluentui-v9-icons';
import { NamedFC } from 'common/react/named-fc';
import * as React from 'react';
import styles from './insights-command-button.scss';

React.useEffect;
export type InsightsCommandButtonProps = ButtonProps & {
    iconName?: any;
    componentRef?: any;
    text?: any;
    menuIconProps?: any;
};

// See https://www.figma.com/file/Wj4Ggf6GGQBQkiDIaHfXRX2B/Accessibility-Insights%3A-Styles?node-id=1%3A27

export const InsightsCommandButton = NamedFC<InsightsCommandButtonProps>(
    'InsightsCommandButton',
    props => {
        const iconComponent = props.iconName;
        const menuIconComponent = props.menuIconProps ? props.menuIconProps.iconName : null;

        return (
            <>
                <h3>InsightsCommandButton</h3>
                <Button
                    appearance="transparent"
                    {...props}
                    icon={Icons[iconComponent]}
                    className={css(styles.insightsCommandButton, props.className)}
                    ref={props.componentRef}
                >
                    {props.text}
                    {menuIconComponent !== null && Icons[menuIconComponent]}
                </Button>
            </>
        );
    },
);
