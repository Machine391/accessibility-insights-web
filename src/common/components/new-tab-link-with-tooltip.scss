// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
@import '../../common/styles/colors.scss';

:global(.high-contrast-theme) {
    .insights-tooltip {
        outline-style: solid;
        outline-width: 1px;
        outline-color: $always-white;
    }
}

.insights-tooltip-host {
    display: inline-block;
    min-width: 0;
}

.insights-link {
    display: flex;

    &:hover,
    &:active {
        text-decoration: none !important;
    }

    &:focus-visible {
        box-shadow: $neutral-60 0 0 0 1px inset !important;
        outline: $neutral-60 auto 1px !important;
        text-decoration: none !important;
    }
}
