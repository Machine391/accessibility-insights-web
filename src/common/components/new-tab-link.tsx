// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { Link, LinkProps, mergeClasses } from '@fluentui/react-components';
import * as React from 'react';

import { NamedFC } from '../react/named-fc';

export type NewTabLinkProps = LinkProps & { as?: 'a' };

export const NewTabLink = NamedFC<NewTabLinkProps>('NewTabLink', ({ className, ...props }) => {
    return <Link className={mergeClasses('insights-link', className)} target="_blank" {...props} />;
});
