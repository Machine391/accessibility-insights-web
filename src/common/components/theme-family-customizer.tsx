// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { ThemeProvider, PartialTheme } from '@fluentui/react';
import { FluentProvider, webDarkTheme, webLightTheme } from '@fluentui/react-components';
import { NamedFC } from 'common/react/named-fc';
import { DefaultTheme } from 'common/styles/default-theme';
import { FastPassTheme } from 'common/styles/fast-pass-theme';
import { FastPassV9Theme } from 'common/styles/fast-pass-v9-theme';
import { HighContrastTheme } from 'common/styles/high-contrast-theme';
import { UserConfigurationStoreData } from 'common/types/store-data/user-configuration-store';
import * as React from 'react';

export type ThemeFamily = 'default' | 'fast-pass';

const themeFamilyDefaultThemes: { [themeFamily in ThemeFamily]: PartialTheme } = {
    default: DefaultTheme,
    'fast-pass': FastPassTheme,
};

const themeFamilyDefaultV9Themes: { [themeFamily in ThemeFamily] } = {
    default: webLightTheme,
    'fast-pass': FastPassV9Theme,
};

export type ThemeFamilyCustomizerProps = {
    userConfigurationStoreData: UserConfigurationStoreData;
    themeFamily: ThemeFamily;
    children: React.ReactNode;
};

// This is a store-state-aware version of the Fluent UI <ThemeProvider> component. It's intended
// as a wrapper for Fluent UI-based components that need to use a different Theme from our default
// one, but still want to support being overriden by HighContrastTheme when a user has enabled our
// High Contrast Mode setting (or a native system High Contrast setting).
export const ThemeFamilyCustomizer = NamedFC<ThemeFamilyCustomizerProps>(
    'ThemeCustomizer',
    props => {
        const isHighContrastEnabled = props.userConfigurationStoreData?.enableHighContrast === true;
        const defaultTheme = themeFamilyDefaultThemes[props.themeFamily];
        const defaultV9Theme = themeFamilyDefaultV9Themes[props.themeFamily]
        const activeTheme = isHighContrastEnabled ? HighContrastTheme : defaultTheme;
        const activeV9Theme = isHighContrastEnabled ? { ...webDarkTheme, colorNeutralBackground1: '#161616 ' } : defaultV9Theme;
        return <ThemeProvider applyTo="body" theme={activeTheme}><FluentProvider theme={activeV9Theme}>{props.children}</FluentProvider></ThemeProvider>;
    },
);
