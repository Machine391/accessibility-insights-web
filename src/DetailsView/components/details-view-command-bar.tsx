// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { NewTabLinkWithTooltip } from 'common/components/new-tab-link-with-tooltip';
import { VisualizationConfigurationFactory } from 'common/configs/visualization-configuration-factory';
import { CardsViewModel } from 'common/types/store-data/card-view-model';
import { FeatureFlagStoreData } from 'common/types/store-data/feature-flag-store-data';
import { ScanMetadata } from 'common/types/store-data/unified-data-interface';
import { UserConfigurationStoreData } from 'common/types/store-data/user-configuration-store';
import { TabStopRequirementState } from 'common/types/store-data/visualization-scan-result-data';
import { VisualizationStoreData } from 'common/types/store-data/visualization-store-data';
import { VersionedAssessmentData } from 'common/types/versioned-assessment-data';
import { VisualizationType } from 'common/types/visualization-type';
import { DetailsViewActionMessageCreator } from 'DetailsView/actions/details-view-action-message-creator';
import { CommandBarButtonsMenu } from 'DetailsView/components/command-bar-buttons-menu';
import styles from 'DetailsView/components/details-view-command-bar.scss';
import { DetailsViewSwitcherNavConfiguration } from 'DetailsView/components/details-view-switcher-nav';
import { InvalidLoadAssessmentDialog } from 'DetailsView/components/invalid-load-assessment-dialog';
import {
    LoadAssessmentButtonDeps,
    LoadAssessmentButtonProps,
} from 'DetailsView/components/load-assessment-button';
import {
    LoadAssessmentDialog,
    LoadAssessmentDialogDeps,
} from 'DetailsView/components/load-assessment-dialog';
import { NarrowModeStatus } from 'DetailsView/components/narrow-mode-detector';
import { QuickAssessToAssessmentDialog } from 'DetailsView/components/quick-assess-to-assessment-dialog';
import { ReportExportButton } from 'DetailsView/components/report-export-button';
import {
    ReportExportDialogFactoryDeps,
    ReportExportDialogFactoryProps,
} from 'DetailsView/components/report-export-dialog-factory';
import {
    SaveAssessmentButtonFactoryDeps,
    SaveAssessmentButtonFactoryProps,
} from 'DetailsView/components/save-assessment-button-factory';
import { ShouldShowReportExportButtonProps } from 'DetailsView/components/should-show-report-export-button';
import { StartOverFactoryDeps } from 'DetailsView/components/start-over-component-factory';
import {
    dialogClosedState,
    StartOverDialog,
    StartOverDialogDeps,
    StartOverDialogProps,
    StartOverDialogState,
    StartOverDialogType,
} from 'DetailsView/components/start-over-dialog';
import {
    TransferToAssessmentButtonDeps,
    TransferToAssessmentButtonProps,
} from 'DetailsView/components/transfer-to-assessment-button';
import { DataTransferViewStoreData } from 'DetailsView/data-transfer-view-store';
import * as React from 'react';
import { ReportGenerator } from 'reports/report-generator';
import { AssessmentStoreData } from '../../common/types/store-data/assessment-result-data';
import { TabStoreData } from '../../common/types/store-data/tab-store-data';
import { DetailsRightPanelConfiguration } from './details-view-right-panel';
import { RefObject } from 'react';

export type ButtonRefFunction = (ref: RefObject<HTMLButtonElement>) => void;

export type DetailsViewCommandBarDeps = {
    getCurrentDate: () => Date;
    reportGenerator: ReportGenerator;
    getDateFromTimestamp: (timestamp: string) => Date;
    detailsViewActionMessageCreator: DetailsViewActionMessageCreator;
} & SaveAssessmentButtonFactoryDeps &
    LoadAssessmentButtonDeps &
    TransferToAssessmentButtonDeps &
    StartOverFactoryDeps &
    LoadAssessmentDialogDeps &
    ReportExportDialogFactoryDeps &
    StartOverDialogDeps;

export type CommandBarProps = DetailsViewCommandBarProps;

export type DetailsViewCommandBarState = {
    isInvalidLoadAssessmentDialogOpen: boolean;
    isLoadAssessmentDialogOpen: boolean;
    isReportExportDialogOpen: boolean;
    loadedAssessmentData: VersionedAssessmentData;
    startOverDialogState: StartOverDialogState;
};

export type ReportExportDialogFactory = (
    props: ReportExportDialogFactoryProps,
) => JSX.Element | null;

export type SaveAssessmentButtonFactory = (
    props: SaveAssessmentButtonFactoryProps,
) => JSX.Element | null;
export type LoadAssessmentButtonFactory = (props: LoadAssessmentButtonProps) => JSX.Element | null;
export type TransferToAssessmentButtonFactory = (
    props: TransferToAssessmentButtonProps,
) => JSX.Element | null;

export interface DetailsViewCommandBarProps {
    deps: DetailsViewCommandBarDeps;
    tabStoreData: TabStoreData;
    assessmentStoreData: AssessmentStoreData;
    rightPanelConfiguration: DetailsRightPanelConfiguration;
    visualizationStoreData: VisualizationStoreData;
    automatedChecksCardsViewData: CardsViewModel;
    needsReviewCardsViewData: CardsViewModel;
    switcherNavConfiguration: DetailsViewSwitcherNavConfiguration;
    scanMetadata: ScanMetadata;
    narrowModeStatus: NarrowModeStatus;
    visualizationConfigurationFactory: VisualizationConfigurationFactory;
    selectedTest: VisualizationType;
    tabStopRequirementData: TabStopRequirementState;
    userConfigurationStoreData: UserConfigurationStoreData;
    featureFlagStoreData: FeatureFlagStoreData;
    dataTransferViewStoreData: DataTransferViewStoreData;
}
export const DetailsViewCommandBar = props => {
    //export const DetailsViewCommandBar = React.forwardRef<HTMLButtonElement, DetailsViewCommandBarProps>((props, ref) => {
    let exportDialogCloseFocus: React.RefObject<HTMLButtonElement> =
        React.useRef<HTMLButtonElement>(null);
    let startOverDialogCloseFocus: React.RefObject<HTMLButtonElement> =
        React.useRef<HTMLButtonElement>(null);
    let transferToAssessmentDialogCloseFocus: React.RefObject<HTMLButtonElement> =
        React.useRef<HTMLButtonElement>(null);

    const [localState, setLocalState]: any = React.useState({
        isInvalidLoadAssessmentDialogOpen: false,
        isLoadAssessmentDialogOpen: false,
        isReportExportDialogOpen: false,
        loadedAssessmentData: {} as VersionedAssessmentData,
        startOverDialogState: 'none',
    });

    const renderTargetPageInfo = (): JSX.Element => {
        const targetPageTitle: string | undefined = props?.scanMetadata.targetAppInfo.name;
        const tooltipContent = `Switch to target page: ${targetPageTitle}`;
        return (
            <div className={styles.detailsViewTargetPage} aria-labelledby="switch-to-target">
                <span id="switch-to-target">Target page:&nbsp;</span>
                <NewTabLinkWithTooltip
                    tooltipContent={tooltipContent}
                    role="link"
                    className={styles.targetPageLink}
                    onClick={props?.deps.detailsViewActionMessageCreator.switchToTargetTab}
                    aria-label={tooltipContent}
                >
                    <span className={styles.targetPageTitle}>{targetPageTitle}</span>
                </NewTabLinkWithTooltip>
            </div>
        );
    };

    const renderFarItems = (): JSX.Element | null => {
        if (props?.narrowModeStatus.isCommandBarCollapsed) {
            return renderCommandButtonsMenu();
        } else {
            return renderCommandButtons();
        }
    };

    const renderCommandButtons = (): JSX.Element | null => {
        const reportExportElement: JSX.Element | null = renderExportButton();
        const startOverElement: JSX.Element = renderStartOverButton();
        const saveAssessmentButtonElement: JSX.Element | null = renderSaveAssessmentButton();
        const loadAssessmentButtonElement: JSX.Element | null = renderLoadAssessmentButton();
        const transferToAssessmentButtonElement: JSX.Element | null =
            renderTransferToAssessmentButton();

        if (
            reportExportElement ||
            saveAssessmentButtonElement ||
            loadAssessmentButtonElement ||
            transferToAssessmentButtonElement ||
            startOverElement
        ) {
            return (
                <div className={styles.detailsViewCommandButtons}>
                    {reportExportElement}
                    {saveAssessmentButtonElement}
                    {loadAssessmentButtonElement}
                    {transferToAssessmentButtonElement}
                    {startOverElement}
                </div>
            );
        }

        return null;
    };

    const renderCommandButtonsMenu = (): JSX.Element => {
        return (
            <CommandBarButtonsMenu
                renderExportReportButton={renderExportButton}
                saveAssessmentButton={renderSaveAssessmentButton()}
                loadAssessmentButton={renderLoadAssessmentButton()}
                transferToAssessmentButton={renderTransferToAssessmentButton()}
                getStartOverMenuItem={getStartOverMenuItem}
                //buttonRef={(ref: React.RefObject<HTMLButtonElement>) => startOverDialogCloseFocus = ref ?? undefined}
                // buttonRef={ref => {
                //     exportDialogCloseFocus = ref ?? undefined;
                //     startOverDialogCloseFocus = ref ?? undefined;
                //     transferToAssessmentDialogCloseFocus = ref ?? undefined;
                // }}
                // ref={ref => {
                //     startOverDialogCloseFocus.current = ref ?? undefined;
                //     exportDialogCloseFocus.current = ref ?? undefined;
                //     transferToAssessmentDialogCloseFocus.current = ref ?? undefined;
                // }
                //ref={startOverDialogCloseFocus}
                // buttonRef={ref => {
                //     startOverDialogCloseFocus = ref ?? undefined;
                //     exportDialogCloseFocus = ref ?? undefined;
                //     transferToAssessmentDialogCloseFocus = ref ?? undefined;
                // }}
                buttonRef={ref => {
                    startOverDialogCloseFocus = ref ?? undefined;
                    exportDialogCloseFocus = ref ?? undefined;
                    transferToAssessmentDialogCloseFocus = ref ?? undefined;
                }}
                // ref={ref => {
                //     startOverDialogCloseFocus = ref ?? undefined;
                //     exportDialogCloseFocus = ref ?? undefined;
                //     transferToAssessmentDialogCloseFocus = ref ?? undefined;
                // }}
            />
        );
    };

    const showReportExportDialog = () => setLocalState({ isReportExportDialogOpen: true });

    const dismissReportExportDialog = () => setLocalState({ isReportExportDialogOpen: false });

    const focusReportExportButton = () => exportDialogCloseFocus?.current?.focus();

    const focusTransferToAssessmentButton = () =>
        transferToAssessmentDialogCloseFocus?.current?.focus();

    const renderExportButton = () => {
        const shouldShowReportExportButtonProps: ShouldShowReportExportButtonProps = {
            visualizationConfigurationFactory: props?.visualizationConfigurationFactory,
            selectedTest: props?.selectedTest,
            tabStoreData: props?.tabStoreData,
        };

        const showButton = props?.switcherNavConfiguration.shouldShowReportExportButton(
            shouldShowReportExportButtonProps,
        );

        if (!showButton) {
            return null;
        }
        return (
            <ReportExportButton
                showReportExportDialog={showReportExportDialog}
                //buttonRef={ref => (exportDialogCloseFocus = ref ?? undefined)}
                buttonRef={exportDialogCloseFocus}
            />
        );
    };

    const renderExportDialog = (): JSX.Element | null => {
        return props?.switcherNavConfiguration.ReportExportDialogFactory({
            ...props,
            isOpen: localState.isReportExportDialogOpen,
            dismissExportDialog: dismissReportExportDialog,
            afterDialogDismissed: focusReportExportButton,
        });
    };

    const renderSaveAssessmentButton = (): JSX.Element | null => {
        return props?.switcherNavConfiguration.SaveAssessmentButton({
            ...props,
            ...{
                buttonRef: transferToAssessmentDialogCloseFocus,
            },
        });
    };

    const renderTransferToAssessmentButton = (): JSX.Element | null => {
        return props?.switcherNavConfiguration.TransferToAssessmentButton({
            ...props,
            ...{
                buttonRef: transferToAssessmentDialogCloseFocus,
            },
        });
    };

    const renderTransferToAssessmentDialog = (): JSX.Element => {
        return (
            <QuickAssessToAssessmentDialog
                isShown={props?.dataTransferViewStoreData.showQuickAssessToAssessmentConfirmDialog}
                afterDialogDismissed={focusTransferToAssessmentButton}
                {...props}
            />
        );
    };

    const renderLoadAssessmentButton = (): JSX.Element | null => {
        return props?.switcherNavConfiguration.LoadAssessmentButton({
            ...props,
            handleLoadAssessmentButtonClick: handleLoadAssessmentButtonClick,
            ...{ buttonRef: transferToAssessmentDialogCloseFocus },
        });
    };

    const renderLoadAssessmentDialog = (): JSX.Element => {
        return (
            <LoadAssessmentDialog
                {...props}
                isOpen={localState.isLoadAssessmentDialogOpen}
                prevTab={props?.assessmentStoreData.persistedTabInfo}
                loadedAssessmentData={localState.loadedAssessmentData}
                tabId={props?.tabStoreData.id!}
                onClose={toggleLoadAssessmentDialog}
            />
        );
    };

    const renderInvalidLoadAssessmentDialog = (): JSX.Element => {
        return (
            <InvalidLoadAssessmentDialog
                {...props}
                isOpen={localState.isInvalidLoadAssessmentDialogOpen}
                onClose={toggleInvalidLoadAssessmentDialog}
            />
        );
    };

    const toggleInvalidLoadAssessmentDialog = () => {
        setLocalState(prevState => ({
            isInvalidLoadAssessmentDialogOpen: !prevState.isInvalidLoadAssessmentDialogOpen,
        }));
    };

    const toggleLoadAssessmentDialog = () => {
        setLocalState(prevState => ({
            isLoadAssessmentDialogOpen: !prevState.isLoadAssessmentDialogOpen,
        }));
    };

    const setAssessmentState = (parsedAssessmentData: VersionedAssessmentData) => {
        setLocalState(_ => ({
            loadedAssessmentData: parsedAssessmentData,
        }));
    };

    const handleLoadAssessmentButtonClick = () => {
        props?.deps.loadAssessmentHelper.getAssessmentForLoad(
            setAssessmentState,
            toggleInvalidLoadAssessmentDialog,
            toggleLoadAssessmentDialog,
            props?.assessmentStoreData.persistedTabInfo,
            props?.tabStoreData.id!,
        );
    };

    const showStartOverDialog = (dialogState: StartOverDialogType) => {
        setLocalState({ startOverDialogState: dialogState });
    };

    const dismissStartOverDialog = () => {
        setLocalState({ startOverDialogState: dialogClosedState });
        startOverDialogCloseFocus?.current?.focus();
    };

    const startOverDialogClosed = (state: DetailsViewCommandBarState): boolean => {
        return state.startOverDialogState === 'none';
    };

    // const componentDidUpdate(prevProps, prevState): void {
    //     // Setting focus after closing the Report Export dialog is handled in the
    //     // afterDialogDismissed prop, which is called after the closing animation.
    //     // Since the start over dialog does not play the closing animation (due
    //     // to flickering issues), we set focus here instead.
    //     if (startOverDialogClosed(localState) && !startOverDialogClosed(prevState)) {
    //         startOverDialogCloseFocus?.focus();
    //     }
    // }

    React.useEffect(() => {
        if (startOverDialogClosed(localState) && localState.startOverDialogState !== 'none') {
            startOverDialogCloseFocus?.current?.removeAttribute('textprediction');
            startOverDialogCloseFocus?.current?.setAttribute('writingsuggestions', 'false');
            startOverDialogCloseFocus?.current?.focus();
        }
    }, [localState.startOverDialogState]);

    const renderStartOverButton = () => {
        const startOverProps = getStartOverProps();
        const startOverComponentFactory = props?.switcherNavConfiguration.StartOverComponentFactory;
        return startOverComponentFactory.getStartOverComponent(startOverProps);
    };

    const getStartOverMenuItem = () => {
        const startOverProps = getStartOverProps('hasSubMenu');
        const startOverComponentFactory = props?.switcherNavConfiguration.StartOverComponentFactory;
        return startOverComponentFactory.getStartOverMenuItem(startOverProps);
    };

    const getStartOverProps = (value?: string) => {
        return {
            ...props,
            // withComponent: true,
            hasSubMenu: value ? true : false,
            openDialog: showStartOverDialog,
            //buttonRef: ref => (startOverDialogCloseFocus = ref),
            buttonRef: startOverDialogCloseFocus,
            // ...{ buttonRef: startOverDialogCloseFocus }
        };
    };

    const renderStartOverDialog = (): JSX.Element => {
        const dialogProps: StartOverDialogProps = {
            ...props,
            dialogState: localState.startOverDialogState,
            dismissDialog: dismissStartOverDialog,
            ...{ buttonRef: startOverDialogCloseFocus },
        };

        return <StartOverDialog {...dialogProps} />;
    };

    return props?.tabStoreData.isClosed ? null : (
        <div className={styles.detailsViewCommandBar} role="region" aria-label="command bar">
            {renderTargetPageInfo()}
            {renderFarItems()}
            {renderExportDialog()}
            {renderInvalidLoadAssessmentDialog()}
            {renderLoadAssessmentDialog()}
            {(localState.startOverDialogState === 'assessment' ||
                localState.startOverDialogState === 'test') &&
                renderStartOverDialog()}
            {renderTransferToAssessmentDialog()}
        </div>
    );
};
