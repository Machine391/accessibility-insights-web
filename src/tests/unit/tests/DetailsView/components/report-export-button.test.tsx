// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { render } from '@testing-library/react';
import { InsightsCommandButton } from 'common/components/controls/insights-command-button';
import { ButtonRefFunction } from 'DetailsView/components/details-view-command-bar';
import {
    ReportExportButton,
    ReportExportButtonProps,
} from 'DetailsView/components/report-export-button';
import * as React from 'react';
import {
    expectMockedComponentPropsToMatchSnapshots,
    mockReactComponents,
    getMockComponentClassPropsForCall,
} from '../../../mock-helpers/mock-module-helpers';

jest.mock('common/components/controls/insights-command-button');

describe(ReportExportButton.displayName, () => {
    mockReactComponents([InsightsCommandButton]);
    let showDialogMock: jest.Mock<() => void> = jest.fn();
    let props: ReportExportButtonProps;

    beforeEach(() => {
        showDialogMock = jest.fn();
        props = {
            showReportExportDialog: showDialogMock,
            buttonRef: {} as ButtonRefFunction,
        };
    });

    it('renders ReportExportButton', () => {
        const renderResult = render(<ReportExportButton {...props} />);
        expect(renderResult.asFragment()).toMatchSnapshot();
        expectMockedComponentPropsToMatchSnapshots([InsightsCommandButton]);
    });

    it('shows export dialog on click', async () => {
        render(<ReportExportButton {...props} />);
        getMockComponentClassPropsForCall(InsightsCommandButton).onClick();
        expect(showDialogMock).toHaveBeenCalledTimes(1);
    });
});
