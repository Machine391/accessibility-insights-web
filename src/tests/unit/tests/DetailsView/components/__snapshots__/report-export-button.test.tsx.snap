// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReportExportButton renders ReportExportButton 1`] = `
<DocumentFragment>
  <mock-insightscommandbutton
    classname="___1qm3wig_0000000 f9hdeju"
    data-automation-id="report-export-button"
    insightscommandbuttoniconprops="[object Object]"
  >
    Export Result
  </mock-insightscommandbutton>
</DocumentFragment>
`;

exports[`ReportExportButton renders ReportExportButton: InsightsCommandButton props 1`] = `
{
  "children": "Export Result",
  "className": "___1qm3wig_0000000 f9hdeju",
  "data-automation-id": "report-export-button",
  "insightsCommandButtonIconProps": {
    "icon": <ArrowExportRegular />,
  },
  "onClick": [MockFunction],
}
`;
